Add the compatibility document for deal.II.

bangerth opened the pull request:
As discussed with Mike Heroux and Lois Curfman McInnes at the SIAM PP conference
last week, we'd like to get deal.II (http://www.dealii.org) into xSDK. This patch
adds the compatibility document. We think that we are pretty clean with regards to
all of these requirements. Going through the list, I've identified a couple of
minor issues that we are currently working on and that all now already have patches.
These are tracked at
  https://github.com/dealii/dealii/issues/6017
  https://github.com/dealii/dealii/issues/6016
Patches relevant to these issues are tracked on the issue pages.

bangerth commented:
@kronbichler, @Rombur, @masterleinad, @davydden, @tjhei, @luca-heltai, @guidokanschat, @tamiko, @jppelteret, @drwells: FYI.

davydden commented on dealii-policy-compatibility.md:
```diff
@@ -0,0 +1,42 @@
+|**M5.** Provide a documented, reliable way to contact the development team. |Full| Mailing list addresses are provided at http://www.dealii.org . |
+|**M6.** Respect system resources and settings made by other previously called packages (e.g. signal handling). |Full| Yes. The only exception that comes to mind is that deal.II enables traps/exceptions when arithmetic operations encounter signaling NaNs. We use signaling NaNs in deal.II to default initialize objects with invalid values to ensure they cannot be used before receiving proper values ...
+|**M7.** Come with an open source (BSD style) license. |Full| LGPL. |
+|**M8.** Provide a runtime API to return the current version number of the software. |Full| Yes: `Utilities::dealii_version_string()`. |
```
you probably don't need `Yes:` here

davydden commented on dealii-policy-compatibility.md:
```diff
@@ -0,0 +1,42 @@
+| Policy                 |Support| Notes                   |
+|------------------------|-------|-------------------------|
+|**R1.** Have a public repository. |Full | http://github.com/dealii/dealii |
+|**R2.** Possible to run test suite under valgrind in order to test for memory corruption issues. |Full| Not at the moment, but we are working on it. |
```
Full->WIP ?

davydden commented on dealii-policy-compatibility.md:
```diff
@@ -0,0 +1,42 @@
+|------------------------|-------|-------------------------|
+|**R1.** Have a public repository. |Full | http://github.com/dealii/dealii |
+|**R2.** Possible to run test suite under valgrind in order to test for memory corruption issues. |Full| Not at the moment, but we are working on it. |
+|**R3.** Adopt and document consistent system for error conditions/exceptions. | Yes. | Yes, we have a system of macros that is uniformly used for this. Errors either lead to termination of the program if, for example, a function is called with invalid arguments (but this can be switched to reporting C++ exceptions). All other error conditions lead to throwing C++ exceptions. |
```
Yes. -> Yes

drwells commented on dealii-policy-compatibility.md:
```diff
@@ -0,0 +1,42 @@
+|**M6.** Respect system resources and settings made by other previously called packages (e.g. signal handling). |Full| Yes. The only exception that comes to mind is that deal.II enables traps/exceptions when arithmetic operations encounter signaling NaNs. We use signaling NaNs in deal.II to default initialize objects with invalid values to ensure they cannot be used before receiving proper values ...
+|**M7.** Come with an open source (BSD style) license. |Full| LGPL. |
+|**M8.** Provide a runtime API to return the current version number of the software. |Full| Yes: `Utilities::dealii_version_string()`. |
+|**M9.** Use a limited and well-defined symbol, macro, library, and include file name space. |Full| All header files are prefixed with `deal.II/`. All C++ names are in a namespace `dealii::`. All macros are prefixed with `DEAL_II_`. |
```
It might be worth mentioning that we overload common math functions (`std::pow`, `std::cos`, etc) for `VectorizedArray`. These are in `std::`.

drwells replied:
The statement 'All macros are prefixed with `DEAL_II_` is not true: `Assert`, `AssertThrow`, `DeclException`, etc. are all macros that do not satisfy this condition.

bangerth replied:
Hm. Both comments are correct. I'll update the statement accordingly.

@drwells -- can you think of any other macros, other than the ones listed?

drwells replied:
I think I got all of them, but if we need everything I'll look more carefully:

```
./deal.II/base/exceptions.h:#define Assert(cond, exc)                                                   
./deal.II/base/exceptions.h:#define Assert(cond, exc)                                                   
./deal.II/base/exceptions.h:#define AssertNothrow(cond, exc)                                              
./deal.II/base/exceptions.h:#define AssertNothrow(cond, exc)                                              
./deal.II/base/exceptions.h:#define AssertThrow(cond, exc)                                              
./deal.II/base/exceptions.h:#define AssertThrow(cond, exc)                                              
./deal.II/base/exceptions.h:#define DeclException0(Exception0)                                        
./deal.II/base/exceptions.h:#define DeclExceptionMsg(Exception, defaulttext)                          
./deal.II/base/exceptions.h:#define DeclException1(Exception1, type1, outsequence)                    
./deal.II/base/exceptions.h:#define DeclException2(Exception2, type1, type2, outsequence)             
./deal.II/base/exceptions.h:#define DeclException3(Exception3, type1, type2, type3, outsequence)      
./deal.II/base/exceptions.h:#define DeclException4(Exception4, type1, type2, type3, type4, outsequence) 
./deal.II/base/exceptions.h:#define DeclException5(Exception5, type1, type2, type3, type4, type5, outsequence) 
./deal.II/base/exceptions.h:#define DeclException0(Exception0)                                        
./deal.II/base/exceptions.h:#define DeclExceptionMsg(Exception, defaulttext)                          
./deal.II/base/exceptions.h:#define DeclException1(Exception1, type1, outsequence)                    
./deal.II/base/exceptions.h:#define DeclException2(Exception2, type1, type2, outsequence)             
./deal.II/base/exceptions.h:#define DeclException3(Exception3, type1, type2, type3, outsequence)      
./deal.II/base/exceptions.h:#define DeclException4(Exception4, type1, type2, type3, type4, outsequence) 
./deal.II/base/exceptions.h:#define DeclException5(Exception5, type1, type2, type3, type4, type5, outsequence) 
./deal.II/base/exceptions.h:#define AssertDimension(dim1,dim2) 
./deal.II/base/exceptions.h:#define AssertVectorVectorDimension(vec,dim1,dim2) 
./deal.II/base/exceptions.h:#define AssertIndexRange(index,range) 
./deal.II/base/exceptions.h:#define AssertIsFinite(number) 
./deal.II/base/exceptions.h:#define AssertThrowMPI(error_code) 
./deal.II/base/exceptions.h:#define AssertThrowMPI(error_code)
./deal.II/base/exceptions.h:#define AssertCuda(error_code) 
./deal.II/base/exceptions.h:#define AssertCusparse(error_code) 
./deal.II/sundials/arkode.h:#define AssertARKode(code) 
./deal.II/sundials/ida.h:#define AssertIDA(code) 
./deal.II/sundials/kinsol.h:#define AssertKINSOL(code) 
./deal.II/matrix_free/cuda_matrix_free.templates.h:#define BLOCK_SIZE 
./deal.II/matrix_free/cuda_matrix_free.templates.h:#define MAX_ELEM_DEGREE
```

jppelteret replied:
> It might be worth mentioning that we overload common math functions (std::pow, std::cos, etc) for VectorizedArray. These are in std::

Ditto for Adol-C ([`adolc_math.h`](https://github.com/dealii/dealii/blob/master/include/deal.II/differentiation/ad/adolc_math.h)) and some other customised functions for Sacado (['sacado_math.h'](https://github.com/dealii/dealii/blob/master/include/deal.II/differentiation/ad/sacado_math.h))

bangerth replied:
I'm not worried about the `std::pow(...)` overloads -- they can't lead to ambiguities.

About `BLOCK_SIZE` and `MAX_ELEM_DEGREE`: @Rombur and @kronbichler, could these be renamed to something with the `DEAL_II_` prefix? Or could we actually make them into `constexpr` constants rather than preprocessor defines?

drwells replied:
We had an issue on the mailing list recently where a user wanted to do a matrix free projection with extra blocks. We should just put these types of things inside `numbers`.

kronbichler replied:
> About `BLOCK_SIZE` and `MAX_ELEM_DEGREE`

We should definitely change them. I think `static constexpr` variables should be fine for both. For `BLOCK_SIZE` it is immediately clear, for the latter it is used for an array to be declared a line below, so that should work as well. But I let someone with regular access to a GPU make the change - for me it would take a few days until I get a booking where I can compile with CUDA.

Rombur replied:
I'll fix it. `BLOCK_SIZE` is an oversight, we were using this macros in several places and I thought I removed all of them but I guess I forgot one. I'll change `MAX_ELEM_DEGREE` too.

@drwells `BLOCK_SIZE` has nothing to do with `BlockVector` it is used to define how you lunch a kernel.

bangerth commented:
I've pushed an updated version.

drwells commented on dealii-policy-compatibility.md:
```diff
@@ -0,0 +1,43 @@
+|**R2.** Possible to run test suite under valgrind in order to test
+|for memory corruption issues. | Work in progress | Not at the moment, but we are working on it. |
+|**R3.** Adopt and document consistent system for error conditions/exceptions. | Yes | Yes, we have a system of macros that is uniformly used for this. Errors either lead to termination of the program if, for example, a function is called with invalid arguments (but this can be switched to reporting C++ exceptions). All other error conditions lead to throwing C++ exceptions. |
+|**R4.** Free all system resources acquired as soon as they are no longer needed. |Full| Yes, that's the goal. |
```
Does this account for `VectorMemory` and friends? I would argue that we do not release system resources since we keep vectors around, though this point is pedantic.

bangerth replied:
Hm, good question. What should we do about it? We keep these vectors around because we typically do need them again.

drwells replied:
I think it is worth mentioning that `GrowingVectorMemory<VectorType>` caches temporary vectors between calls to a Krylov solver.

bangerth replied:
I've updated that as well.

davydden commented on dealii-policy-compatibility.md:
```diff
@@ -0,0 +1,43 @@
+|**M13.** Install headers and libraries under \<prefix\>/include and \<prefix\>/lib. |Full| Yes. The install path is chosen via `CMAKE_INSTALL_PREFIX`. |
+|**M14.** Be buildable using 64 bit pointers. 32 bit is optional. |Full| We make no assumptions. Whatever the compiler wants to use. |
+|**M15.** All xSDK compatibility changes should be sustainable. |Full| Yes. Rather than providing the xSDK options as add ons, we have incorporated them fully into the basic deal.II infrastructure, so that xSDK options get the same maintaince and support as any other feature. |
+|**M16.** The package must support production-quality installation compatible with the xSDK install tool and xSDK metapackage. |Full| Yes. deal.II is supported by SPACK. |
```
since this mentions `xSDK metapackage`, i guess we need to add `deal.II` to https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/xsdk/package.py

If that's the case, I can open a PR for that.

bangerth replied:
Yes, go ahead. We may be getting ahead of ourselves, but I see no reason not to open such a PR already at this time even while we're figuring other things out.

davydden replied:
@bangerth 

> Yes, go ahead. We may be getting ahead of ourselves, but I see no reason not to open such a PR already at this time even while we're figuring other things out.

see https://github.com/spack/spack/pull/7529

bangerth commented:
On 03/12/2018 06:42 AM, Bruno Turcksin wrote:
> @drwells <https://github.com/drwells> |BLOCK_SIZE| has nothing to do with 
> |BlockVector| it is used to define how you lunch a kernel.

Is lunching a kernel related to having popcorn around mid-day?

(But seriously: thanks for offering to address these points.)

jwillenbring commented:
The only issue I saw was a minor one with R2 being broken into 2 rows. I tried to submit a pull request, but there was an error, maybe because I was trying to submit a pull reuqest to a fork. Are you ready for this PR to be merged?

masterleinad commented on dealii-policy-compatibility.md:
```diff
@@ -0,0 +1,43 @@
+|------------------------|-------|-------------------------|
+|**R1.** Have a public repository. |Full | http://github.com/dealii/dealii |
+|**R2.** Possible to run test suite under valgrind in order to test
+|for memory corruption issues. | Work in progress | Not at the moment, but we are working on it. |
```
This was addressed in dealii/dealii#6035.

bangerth commented:
@jwillenbring -- I fixed the awkward wrap for R2.
@masterleinad -- I also updated R2 for the valgrind patch.

I think this is now in final form and ready for merge!

bangerth commented:
@jwillenbring -- I think this is ready from our side to be merged. Anything else you want me to do?

bangerth commented:
Ping?

bangerth commented:
No worries, @jwillenbring. What are the next steps for us to take in this process?
